cmd/lncli/cmd_pay.go

```diff
@@ -458,23 +464,199 @@ func trackPayment(ctx *cli.Context) error {
 	}
 
-	stream, err := client.TrackPayment(context.Background(), req)
+	stream, err := routerClient.TrackPayment(context.Background(), req)
 	if err != nil {
 		return err
 	}
 
+	client := lnrpc.NewLightningClient(conn)
+	_, err = printLivePayment(stream, client, ctx.Bool(jsonFlag.Name))
+	return err
+}
+
+// printLivePayment receives payment updates from the given stream and either
+// outputs them as json or as a more user-friendly formatted table. The table
+// option uses terminal control codes to rewrite the output. This call
+// terminates when the payment reaches a final state.
+func printLivePayment(stream routerrpc.Router_TrackPaymentClient,
+	client lnrpc.LightningClient, json bool) (*lnrpc.Payment, error) {
+
+	// Initialize terminal (re)writer if formatted output is requested.
+	var writer *uilive.Writer
+	if !json {
+		writer = uilive.New()
+		writer.Start()
+		defer writer.Stop()
+	}
+
+	aliases := newAliasCache(client)
+
+	first := true
 	for {
-		status, err := stream.Recv()
+		payment, err := stream.Recv()
 		if err != nil {
-			return err
+			return nil, err
+		}
+
+		if json {
+			// Delimit json messages by newlines (inspired by
+			// grpc over rest chunking).
+			if first {
+				first = false
+			} else {
+				fmt.Println()
+			}
+
+			// Write raw json to stdout.
+			printRespJSON(payment)
+		} else {
+			// Write formatted table to writer. This writer will
+			// erase the previously written output in the terminal
+			// first.
+			table := formatPayment(payment, aliases)
+			_, err := writer.Write([]byte(table))
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Terminate loop if payments state is final.
+		if payment.Status != lnrpc.Payment_IN_FLIGHT {
+			return payment, nil
 		}
+	}
+}
 
-		printRespJSON(status)
+// aliasCache allows cached retrieval of node aliases.
+type aliasCache struct {
+	cache  map[string]string
+	client lnrpc.LightningClient
+}
 
-		if status.Status != lnrpc.Payment_IN_FLIGHT {
-			return nil
+func newAliasCache(client lnrpc.LightningClient) *aliasCache {
+	return &aliasCache{
+		client: client,
+		cache:  make(map[string]string),
+	}
+}
+
+// get returns a node alias either from cache or freshly requested from lnd.
+func (a *aliasCache) get(pubkey string) string {
+	alias, ok := a.cache[pubkey]
+	if ok {
+		return alias
+	}
+
+	// Request node info.
+	resp, err := a.client.GetNodeInfo(
+		context.Background(),
+		&lnrpc.NodeInfoRequest{
+			PubKey: pubkey,
+		},
+	)
+	if err != nil {
+		// If no info is available, use the
+		// pubkey as identifier.
+		alias = pubkey[:6]
+	} else {
+		alias = resp.Node.Alias
+	}
+	a.cache[pubkey] = alias
+
+	return alias
+}
+
+// formatMsat formats msat amounts as fractional sats.
+func formatMsat(amt int64) string {
+	return strconv.FormatFloat(float64(amt)/1000.0, 'f', -1, 64)
+}
+
+// formatPayment formats the payment state as an ascii table.
+func formatPayment(payment *lnrpc.Payment, aliases *aliasCache) string {
+	t := table.NewWriter()
```

Comment:
personal preference: the `table.StyleLight` looked the cleanest to me

Reply:
I tried that, but it messed up my terminal

Reply:
weird, i tried it locally and it worked for me. oh well

Reply:
I don't know what the reason is that it doesn't work. Maybe for the next release we can test a bit wider on different OSes. Who knows how Windows renders the fancy table?